### DISCUSSION TITLE
I'd like to pay for the creation of a simple Stellarium v19.3 script (macOS)

### DISCUSSION THREAD

**Discussion by ImmersiveTheatres** [Stellarium/stellarium#3424] [Q&A]
Hello Stellarium community

For various reasons, I am currently restricted to using Stellarium v19.3 in my mobile planetariums (immersive-theatres.com).

Stellarium v19.3 (macOS) works very well for me - and as I'm sure you know, v19.3 includes many scripts that work very well too.

I now need to create a simple automated script to execute a few simple diurnal and annual motion commands.

I am very happy to pay for this, as I know there are not many people with the expertise to create scripts for v19.3.

**If you are available to help, please [email](mailto:mario@immersive-theatres.com) or DM me, so I can share my requirements with you, and you can quote me a price**.

Thanks very much

Regards

Mario

**Comment by alex-w** [comment]
> I now need to create a simple automated script to execute a few simple diurnal and annual motion commands.

Please more details - which motions you need to see?

**Comment by ImmersiveTheatres** [comment]
Hello Alexander

Thank you for your response.

The simple automated script I require is for a Christams Star planetarium
show I am preparing.

   1. begin with current time and location (Bristol, UK); allowing manual
   movement of sky as normal

   2. then I launch the Christmas Star script (automatically pausing after
   each step, advancing with spacebar):


   - move from present time & location to Bethlehem (31.7° N / 35.2° E)
   before sunrise on 12 August 3 BC
   - zodiac constellations appear, with planet labels
   - diurnal motion forward a little, for a better view of Jupiter and Leo;
   calendar date appears at the front
   - annual motion forward until morning of 01 June 2 BC, all
   constellations fade except Leo
   - diurnal motion forward to evening of 01 June 2 BC
   - annual motion forward until 17 June 2 BC
   - all constellations appear, slow diurnal motion


That's it.

Please let me know the cost for this. I would like to pay, because you guys
do so much for free, I'd like to reward your skills with a financial
contribution.

Thanks very much

Regards

Mario

**Comment by ImmersiveTheatres** [comment]
Hello Alexander / Stellarium Community

**I need to know please if it will be possible to create the v19.3 script (macOS) described above?** If not, I'm going to have to search for a different planetarium application for my Christmas show later this year.

I have two related questions that may help:

- if the scripting syntax has not changed between v19.3 and v22.3, then why does my v19.3 play the default v19.3 scripts fine, but it does not play [this script](https://github.com/hn-88/stellarium-scripts/blob/main/annual-motion-live.ssc)?

- I know someone who has offered to create my required script for v22.3. If I ask him to proceed, will it be possible to convert the v22.3 script into a v19.3 script? 

**As I said, I'm very happy to pay for help, as this is becoming quite urgent for me now.**

Thanks very much

Regards

Mario

**Reply by gzotti** [reply]
You script from obsolete 0.19.* should at least play on 23.3-Qt5. Differences between Qt5 and Qt6-based scripting are described in the User Guide. 

Sorry, we cannot give much support for obsolete versions. "Does not play" is a bit vague. Are there errors in the logfile? Does it crash? It looks simple enough. Commands/syntax/arguments are still documented under http://stellarium.org/doc/0.19/

**Reply by alex-w** [reply]
> I need to know please if it will be possible to create the v19.3 script (macOS) described above? If not, I'm going to have to search for a different planetarium application for my Christmas show later this year.

Yes, it's possible, but without advancing with spacebar (other shortcut by technical reasons).

> if the scripting syntax has not changed between v19.3 and v22.3, then why does my v19.3 play the default v19.3 scripts fine, but it does not play [this script](https://github.com/hn-88/stellarium-scripts/blob/main/annual-motion-live.ssc)?

The scripting language a bit changes plus it has a bit different behaviour for Qt5- and Qt6-based packages. A simple scripts will run identically. The selected script has typos (missing `;`) - at least it run in 23.3-Qt6 after typofixes.

> I know someone who has offered to create my required script for v22.3. If I ask him to proceed, will it be possible to convert the v22.3 script into a v19.3 script?

I think it possible to write a script for both versions :)

**Reply by hn-88** [reply]
Thanks @alex-w  , I have added semicolons to [this script](https://github.com/hn-88/stellarium-scripts/blob/main/annual-motion-live.ssc) now.

**Comment by ImmersiveTheatres** [comment]
Thank you for your reply Georg & Alexander.

I would LOVE to delve deeper into the scripting and learn how to do / fix all this myself, but I really cannot afford the time.

Can you refer my request to a Stellarium community member who (a) would have the time to do this, and (b) would like to get paid for their effort?

Or can one of you please help me achieve this script? If you don't wish to get 'paid', I'll happily make a financial contribution to the Stellarium project, etc.

I just need that simple script to run on v19.3 (macOS)

**Reply by alex-w** [reply]
I can try to write a script in next 2-3 days

**Comment by ImmersiveTheatres** [comment]
Wow. Awesome. THANK YOU Alexander. That is such a relief 😊

No problem about the Spacebar advance of course, any key will do.

Please feel free to [email](mailto:mario@immersive-theatres.com) directly if you wish.

Thank you again 🙏🏻

**Reply by alex-w** [reply]
Please see a sketch of script: https://github.com/alex-w/scripts/blob/master/bethlehem.ssc

No advancing by shortcut yet

**Comment by ImmersiveTheatres** [comment]
That looks great. I'll download it and will try to run in in the next hour and let you know. Thank you!

**Comment by ImmersiveTheatres** [comment]
WOW 🤩🤩 The script is fantastic Alexander! 

It looks and runs great on my macOS (v12.6) flat-screen desktop with Stellarium v22.2

I tried running it on my planetarium / dual-screen macOS (v10.15) with Stellarium v19.3, and:

- the script starts and moves to Bethlehem, but then all the constellations appear (not just the Zodiac) - and it freezes.
- also the horizon is not level (this is a spherical mirror issue if I recall) - photo below
![IMG_7132](https://github.com/Stellarium/stellarium/assets/59133199/6a97d9dd-d3b9-49c2-a6b7-de34d5ab7f0e)

This is a great start, thank you again!

**Reply by alex-w** [reply]
Please check https://github.com/alex-w/scripts/blob/master/bethlehem_m.ssc in v0.19.3

**Comment by ImmersiveTheatres** [comment]
Looking much better! 

It works fine in v19.3 until the annual motion, then it seems to stick and jump very wildly.

Also, after zooming in to Jupiter & Venus, it does not return to a flat horizon.

I recorded a video of the full script and uploaded it here: 
https://www.dropbox.com/scl/fo/omd43xbmpjjb03q5r3hyl/h?rlkey=fs4n6c205whrew98w6ay1ta34&dl=0

(apologies, I'm not in a planetarium today, so I'm projecting it agains a wall)

**Reply by alex-w** [reply]
Please check https://github.com/alex-w/scripts/blob/master/bethlehem.ssc in both versions - it should be work identical now.

Uncomment `waitNextStep();` lines to use button `]` as stepper between scenes.

**Comment by ImmersiveTheatres** [comment]
Great, thank you for v2 Alexander, and for the option to use the  ]  key to step between scenes.

I haven't tried the steps yet. I thought I would first report on the overall script (you can see the video here: 
https://www.dropbox.com/scl/fo/omd43xbmpjjb03q5r3hyl/h?rlkey=fs4n6c205whrew98w6ay1ta34&dl=0 )

- the jump to Bethlehem is very quick / instant - can this be made slower / smoother?
- the annual motion sequence works smoothly and well - but then it jumps and _repeats_ itself very quickly(?)

Other items:
- is it possible to have the time & date appear at the front somewhere?
- the flat horizon no longer functions after the script initiates. Is this because a spherical mirror setting needs to be added to the script?

These are my current specific spherical mirror / video settings (I'm not sure which one is responsible for a flat horizon):

**[navigation]**
init_fov                      = 180


**[projection]**
flag_use_gl_point_sprite        = false
flip_horz                                         = false
flip_vert                                          = false
type                                                   = ProjectionFisheye
viewport                                          = none
viewportMask                                = none
viewport_center_offset_x        = 0
viewport_center_offset_y        = 0


**[video]**
fullscreen                         = true
screen_w                          = 1024
screen_h                          = 768
screen_x                           = 237
screen_y                          = 85
viewport_effect              = sphericMirrorDistorter


**[spheric_mirror]**
distorter_max_fov                                    = 180
dome_radius                                               = 5.0 
flag_use_ext_framebuffer_object     = false
flip_horz                                                       = false
flip_vert                                                         = false
image_distance_div_height                 = 2.67
mirror_position_x                      = 0
mirror_position_y                      = 5
mirror_position_z                      = 0
mirror_radius                                = 0.37
projector_alpha                        = 0
projector_delta                        = 10.6
projector_gamma                        = 0.3
projector_phi                                 = 0
projector_position_x                   = 0
projector_position_y                   = 4
projector_position_z                   = 0
texture_triangle_base_length  = 8



I think this one needs to be applied at the end of the script, because now a toolbar appears:

**[gui]**
flag_show_gui  =  false

**Reply by alex-w** [reply]
Please see updated version of script.

> is it possible to have the time & date appear at the front somewhere?

Yes. Please see `showText()` method - coordinates `20, 20` is a screen coordinates of left top corner of text line.

> the flat horizon no longer functions after the script initiates. Is this because a spherical mirror setting needs to be added to the script?

Probably yes - I've changes something now, please check.

**Comment by ImmersiveTheatres** [comment]
OK thank you Alexander, I have run out of time today, so I will look at it in detail on Sunday and feedback then. 

I have made a third video, as it is not quite right:
https://www.dropbox.com/scl/fo/omd43xbmpjjb03q5r3hyl/h?rlkey=fs4n6c205whrew98w6ay1ta34&dl=0

The date display is important, but as you can see, it does not yet appear on the screen (dome)?

Also, the flat horizon is still not working.

Thank you for all your work so far - this is exciting and I really appreciate it!

**Reply by alex-w** [reply]
Please check the latest changes.

**Comment by ImmersiveTheatres** [comment]
Sorry, video is still uploading

**Comment by ImmersiveTheatres** [comment]
Wow, fantastic Alexander, the horizon issue is now fixed. You can see the fourth video [here](https://www.dropbox.com/scl/fo/omd43xbmpjjb03q5r3hyl/h?rlkey=fs4n6c205whrew98w6ay1ta34&dl=0).

Only two things remain (if possible):

- can the date and time appear just above the South cardinal point, near the horizon? As you can see right now it is appearing high up in the sky in the southeast

- regarding the Annual Motion from 12 August 3 BC to 01 June 2 BC (approx. 10 months): **can we do this with Leo not moving?** In other words, can we keep Leo in the dawn sky, and **simply advance the date one week at a time** until 01 June 2 BC? This way, the only things that will move will be the planets.

Otherwise, thank you, this is looking EXCELLENT, and I'm so pleased that we are accomplishing it in v19.3 ! 😃

**Reply by gzotti** [reply]
You cannot freeze Leo in the Dawn sky when moving time over months. Time of day must change then.

**Reply by alex-w** [reply]
> can the date and time appear just above the South cardinal point, near the horizon? As you can see right now it is appearing high up in the sky in the southeast

Please find `LabelMgr.labelHorizon(text, 0, 45, true, 30, "#ff0000");` in the script and edit `0, 45` values (in degree) to fit the best view in your environment.

**Comment by ImmersiveTheatres** [comment]
Thank you very much for the Label Manager advice, I will position it myself.

**Comment by ImmersiveTheatres** [comment]
Finally, as regards the annual motion, I'm sorry I did not explain it very well. In fact, I actually made a mistake in my original description!

I don't need Leo to remain visible with the annual motion (apologies for this).

What I would like to demonstrate with annual motion (to 01 June 2BC) is the movements of Jupiter and Venus in the East. Especially the retrograde motion of Jupiter.

As long as Jupiter and Venus remain above the eastern horizon as we move time to 01 June 2BC.

(Leo can fade away at the beginning of the annual motion).

If we can please make this final adjustment, the script will be complete (I promise ! 😄)

**Reply by alex-w** [reply]
I fear it's impossible. I can "freeze" the sky and show the movement of planets in same "corner" of sky, but this is possible in equatorial mount type - so, the ground will be moving in this case. But in your use case (a planetarium mode - horizontal mount type) the ground shouldn't be moving.

**Reply by gzotti** [reply]
Yes, this is one difference between optomechanical analog planetarium and the digital systems. The required mode would require a motion along a geographical latitude circle with constant local time, or even adjusting time to keep twilight state constant. What could be done is showing day-to-day state of twilight at constant solar altitude. This is not smooth, but will keep observers in one location.

**Reply by ImmersiveTheatres** [reply]
Yes, thank you for describing what I was unable to explain Georg. 

I am trying to re-create in Stellarium a very successful Christmas show I prepared and delivered 20 years ago when I managed the Glasgow Planetarium (you can read the story [here](https://www.heraldscotland.com/news/12422781.why-christmas-should-perhaps-be-celebrated-on-august-12-scientists-put-up-new-theories-about-star-that-guided-the-magi/)).

With the Zeiss Starmaster projector I simply ran 'annual motion' and the stars remained more-or-less frozen - while the planets continued to move - showing Jupiter's retrograde motion beautifully in the east.

Maybe we could try Georg's suggestion?: _"What could be done is showing day-to-day state of twilight at constant solar altitude. This is not smooth, but will keep observers in one location"_

I simply want the audience to see what Venus & Jupiter were doing every morning between 12 August 3 BC and 01 June 2 BC.

**Comment by ImmersiveTheatres** [comment]
Hello everyone. 

Thank you again for all your help so far with my Christmas Star script. 

I have now found someone knowledgable with Stellarium scripting to help me (a local amateur astronomer); and we are making good progress with the script. We have managed to fine-tune the first three steps to show exactly what I wish to present in the planetarium. 

I don't know how to upload the script file to the Scripts folder, so I have attached it here as a zip file:
[bethlehem_updated.ssc.zip](https://github.com/Stellarium/stellarium/files/13079034/bethlehem_updated.ssc.zip)

YET we have hit a problem. 

We have researched and tried EVERYTHING we can, **but we just cannot work out how to fade out Leo together with the other zodiac constellations in STEP 3**.

In fact, we now suspect there is a bug in v19.3 and that is why we cannot remove Leo.

@alex-w  / @gzotti  / or anyone one else : can you please examine the latest script and see if it is possible to fade out Leo before the annual motion begins in STEP 3 ?

If not, please can you confirm this is because of a bug in v19.3 and there is nothing we can do about it?

Your help with this will be greatly appreciate it, as it is driving us mad! :)

Thank you 🙏🏻

**Reply by gzotti** [reply]
I cannot try specifically with 0.19.3, but just try `ConstellationMgr.setFlagArt(false);`  when you want to disable rendering of constellation art.

**Reply by ImmersiveTheatres** [reply]
🥳

**Reply by ImmersiveTheatres** [reply]
Brilliant! Thank you. We've been using:
ConstellationMgr.deselectConstellations();

Now we know. Thank you again.

Now back to the script .... 🤓

**Comment by ImmersiveTheatres** [comment]
Right, next obstacle encountered (sooner than expected) 😊

With the annual motion in PART 3 of our script, we'd like the date label to update in real-time, as we move the sky from 12 August 3BC to 01 June 2BC.

Below is what we've added to the script - but can't understand why the date label does not change in real-time. It just sits at "12 August 3BC"

======================

//
//  STEP 3 - ALL CONSTELLATIONS FADE, ANNUAL MOTION FORWARD UNTIL MORNING OF 01 JUNE 2 BC.
//


// LabelMgr.deleteAllLabels();
ConstellationMgr.setFlagArt(false);
StelMovementMgr.lookZenith();
core.wait(0.1);

var labelID = LabelMgr.labelScreen("Bethlehem 12 August 3 BC", 160, 10, true, 30, "#ff0000");

function formatDate(dateString) {
    var dateParts = dateString.split("-");
    var year = parseInt(dateParts[0]);
    var month = parseInt(dateParts[1]);
    var day = parseInt(dateParts[2]);

    var monthNames = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
    var formattedMonth = monthNames[month - 1];

    if (year < 0) {
        year = Math.abs(year) + 1 + " BC";  // Convert negative years to BC, adding 1 because there is no year 0
    }

    return day + " " + formattedMonth + " " + year;
}


// real annual motion
for (i=0; i<323; i++)
{
    core.setDate("+1 day");
    var currentDate = core.getDate("utc").toString().slice(0,10);  // Get the current date as a string in the format "YYYY-MM-DD"
    var formattedDate = formatDate(currentDate);  // Format the date to the desired format "DD Month YYYY BC/AD"
    LabelMgr.deleteLabel(labelID);  // Remove the old label
    labelID = LabelMgr.labelScreen("Bethlehem " + formattedDate, 160, 10, true, 30, "#ff0000");  // Create a new label with the updated date and store its ID
    core.wait(0.1);
}


waitNextStep();

**Reply by alex-w** [reply]
See an example: https://github.com/Stellarium/stellarium/blob/2533f48decea2543e2a8a4c8e157f3fc6094acdc/scripts/analemma.ssc#L33

You're missing couple commands

**Comment by ImmersiveTheatres** [comment]
Hi again

Can someone help us please?

**Should I open a new ticket somewhere else - with more people available to help? If so, please can you let me know where?**

We now have a number of other questions - which are probably quite easy to solve for anyone knowledgable with Stellarium scripting - but I'm not sure whether I should post them here (since no-one has addressed my last question).

**Should I be on a larger forum - where more people can see my questions?**

As regards my last question above: can someone just let me know if it IS possible to display the date in real-time during annual motion - or not?

The problem we're experiencing is we don't know if something is possible with Stellarium scripting or not. So we just need some guidance from people who are more experienced.

The Christmas Star script as Alexander has created it is almost perfect. We just need to make some adjustments to it. iow we are very close now, and would like to finish this project.

(and as I've said before, I'm happy to **PAY** someone experienced to do this / help us).

**Reply by gzotti** [reply]
If you find a forum of Stellarium script developers, we'd be interested to know it. We have 500.000 or more users. It seems 100 of them ever use scripting, and 5 discuss it. Sometimes our daytime occupations keep us from answering. We have no department of standby contract developers.

**Reply by ImmersiveTheatres** [reply]
Wow. I didn't realise so few people discuss Stellarium scripting 😲 I guess most of the experienced scriptwriters are working for the commercial versions of Stellarium, like Nightshade and Stellarium360 ?

This is why I began my request by **offering to pay for this script**. 

I understand developers like @gzotti and @alex-w are working for free in their spare time. So I was hoping one of the developers or experienced scripters would charge me a fair price for your expertise & time, to make it worth your while. I really don't mind paying!

My colleague and I appreciate all you help so far, and we are very close to completing the script (we have also learned a LOT in the process, which is very valuable for the future).

But despite all our efforts, we have hit a wall as regards **changing the calendar in real-time during annual motion**. We have studied the analemma example, but can't figure out how to apply the relevant commands to our v19.3 script.

Please can you take a look at the full script below - and specifically at the scripting in STEP 3 : **// Loop to increment the date by two days at a time until "01 June 2 BC"**

The clock just does not run, and we don't know why 😧

Everything else before that works very well.


============================================================

// Name: Christmas Star
// License: Public Domain
// Author: Alexander V. Wolf + Harry + Mario
// For Stellarium version: 0.19.3
// Description: Possible explanation of the Christmas Star over Bethlehem in August 03 BC



// DEFINE THE PAUSE-CONTINUE FUNCTION

function waitNextStep()
{
    // next step button: ]
    var q = 1;
    var jd = core.getJDay();
    while (q==1)
    {
        var jdn = core.getJDay();
        if (jdn-jd>5)
        {
            q=0;
        } else {
            jd = jdn;
        }
        core.wait(0.01);
    }
    core.setJDay(jd);
}




// DEFINE THE SHOW TEXT FUNCTION WITH COLOUR AND POSITION ON THE SCREEN

function showText(text, time)
{
    LabelMgr.labelHorizon(text, 160, 10, true, 30, "#ff0000");
    core.wait(0.1);
}




// DEFINE AN ARRAY OF CONSTELLATIONS THAT CAN BE REFERRED TO AS A GROUP LATER

var zodiac = new Array("Leo", "Cancer", "Gemini", "Taurus", "Aries", "Pisces", "Aquarius");




// DEFINE SPHERICAL MIRROR AND OTHER PARAMETERS

core.setProjectionMode("ProjectionFisheye");
core.wait(0.1); // Wait for 0.1 seconds

// core.setDiskViewport(true); // not sure what this is and why it has been hashed out. Enabling it appears to make no visible difference

core.setFlagGravityLabels(true);
core.setSphericMirror(true);
core.wait(0.1);
core.clear("natural");

//StelMovementMgr.setFlagLockEquPos(true); // this sounds like it's meant to show proper annual motion [see Notes], yet causes havoc when enabled

StelMovementMgr.zoomTo(180, 0.1);
StelMovementMgr.lookZenith();
core.wait(0.1);


// core.setGuiVisible(true); // this enables the control panel bar at the top of the screen



////////////////////////////////////////////////////////////////////////////////////////
//
//  STEP 1 - MOVE (RAPIDLY) FROM CURRENT TIME TO BETHLEHEM (31.7° N / 35.2° E) BEFORE SUNRISE ON 12 AUGUST 3 BC
//


// SET THE NUMBER OD STEPS IN THE TRANSITION

var stepCount = 25;



// OBTAIN THE INITIAL DATE AS A JULIAN DATE NUMBER

var initialJDay = core.getJDay();



// SET THE FINAL DATE AS THE JULIAN DATE NUMBER BEFORE SUNRISE ON 12 AUGUST 3BC 

var finalJDay = 1720551.59164;



// CALCULATE THE INCREMENT FOR EACH STEP

var dateStep = (finalJDay - initialJDay) / stepCount;
var latStep = (31.7 - 51.46) / stepCount;
var longStep = (35.2 + 2.6) / stepCount;


// CALCULATE THE CURRENT DATE, LATITUDE & LONGITUDE FOR THIS STEP

for (var i = 0; i <= stepCount; i++) {

var currentJDay = initialJDay + dateStep * i;
var currentLat = 51.46 + latStep * i;
var currentLong = -2.6 + longStep * i;  // initial longitude is negative (west)


// SET THE CURRENT DATE, LATITUDE AND LONGITUDE

core.setJDay(currentJDay);
core.setObserverLocation(currentLong, currentLat, 1, 5, "Transition Location", "Earth");
core.wait(0.05);  // Adjust this value to control the speed of the transition
}


// SHOW LOCATION, DATE AND CARDINAL POINTS

showText("Bethlehem 12 August 3 BC", 1);
LandscapeMgr.setFlagCardinalsPoints(true);


waitNextStep();



////////////////////////////////////////////////////////////////////////////////////////
//
//  STEP 2 - PLANET LABELS APPEAR, ZODIAC CONSTELLATIONS APPEAR, STAR LABELS APPEAR BRIEFLY AND FADE.
//


// SHOW PLANET LABELS

SolarSystem.setFlagLabels(true);



// WAIT 10 SECONDS

// core.wait(10);  



// SHOW ZODIAC CONSTELLATION ARRAY

ConstellationMgr.deselectConstellations();
ConstellationMgr.setFlagArt(true);
ConstellationMgr.setFlagBoundaries(false);
ConstellationMgr.setFlagLines(false);
ConstellationMgr.setFlagLabels(false);
ConstellationMgr.setFlagIsolateSelected(true);



// HAVE CONSTELLATIONS APPEAR IN REVERSE FROM AQUARIUS TO LEO

for(i = zodiac.length - 1; i >= 0; i--)
{
    core.selectConstellationByName(zodiac[i]);
    core.wait(1.5);
}



// WAIT 10 SECONDS

core.wait(2);  



// SHOW STAR NAMES (UP TO MAGNITUDE 1.4, REGULUS) FOR 5 SECONDS AND FADE

StarMgr.setLabelsAmount(1.4);
StarMgr.setFlagLabels(true);
core.wait(5); 
StarMgr.setFlagLabels(false);



waitNextStep();




////////////////////////////////////////////////////////////////////////////////////////
//
//  STEP 3 - ALL CONSTELLATIONS FADE, ANNUAL MOTION FORWARD UNTIL MORNING OF 12 FEB 2 BC.
//





// FADE CONSTELLATIONS

ConstellationMgr.setFlagArt(false);
StelMovementMgr.lookZenith(); // does this cause the jump? why do we need this?
core.wait(0.1);




waitNextStep();



// TURN ON CONSTELLATION LINES

ConstellationMgr.setFlagLines(true);



waitNextStep();



// TURN ON TRAILS ONLY FOR JUPITER

SolarSystem.setFlagIsolatedTrails(true);
SolarSystem.setFlagTrails(true); 
core.selectObjectByName("Jupiter");




waitNextStep();



// Function to format the date
function formatDate(dateString) {
    var dateParts = dateString.split("-");
    var year = parseInt(dateParts[0]);
    var month = parseInt(dateParts[1]);
    var day = parseInt(dateParts[2]);

    var monthNames = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
    var formattedMonth = monthNames[month - 1];

    if (year < 0) {
        year = Math.abs(year) + 1 + " BC";  // Convert negative years to BC, adding 1 because there is no year 0
    } else {
        year = year + " AD";
    }

    return day + " " + formattedMonth + " " + year;
}




waitNextStep();



// Loop to increment the date by two days at a time until "01 June 2 BC"
while (core.getDate("utc").toString() < "-0001-06-01T00:00:00") {
    core.setDate("+2 days");  // Increment the date by two days
    var currentDate = core.getDate("utc").toString().slice(0,10);  // Get the current date as a string in the format "YYYY-MM-DD"
    var formattedDate = formatDate(currentDate);  // Format the date to the desired format "DD Month YYYY BC/AD"
    LabelMgr.setLabelShow(labelID, false);  // Hide the old label
    labelID = LabelMgr.labelHorizon(formattedDate, 160, 10, true, 30, "#ff0000");  // Create a new label with the updated date and store its ID
    LabelMgr.setLabelShow(labelID, true);  // Show the new label
    core.wait(1);  // Wait for a short duration before proceeding to the next iteration (you can adjust this value to control the speed of time progression)
}






waitNextStep();

=======================================================================

// ORIGINAL / OLD ATTEMPT


// ANNUAL MOTION TO 01 JUNE 2 BC

function formatDate(dateString) {
    var dateParts = dateString.split("-");
    var year = parseInt(dateParts[0]);
    var month = parseInt(dateParts[1]);
    var day = parseInt(dateParts[2]);

    var monthNames = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
    var formattedMonth = monthNames[month - 1];

    if (year < 0) {
        year = Math.abs(year) + 1 + " BC";  // Convert negative years to BC, adding 1 because there is no year 0
    }

    return day + " " + formattedMonth + " " + year;
}




waitNextStep();




// Existing code to create the initial date label
var labelID = LabelMgr.labelScreen("Bethlehem 12 August 3 BC", 160, 10, true, 30, "#ff0000");




waitNextStep();





// real annual motion
for (i=0; i<323; i++)
{
    core.setDate("+1 day");
    var currentDate = core.getDate("utc").toString().slice(0,10); // Get the current date as a string in the format "YYYY-MM-DD"
    var formattedDate = formatDate(currentDate); // Format the date to the desired format "DD Month YYYY BC/AD"
    LabelMgr.setLabelShow(labelID, false); // Hide the old label
    labelID = LabelMgr.labelScreen("Bethlehem " + formattedDate, 160, 10, true, 30, "#ff0000"); // Create a new label with the updated date and store its ID
    LabelMgr.setLabelShow(labelID, true); // Show the new label
    core.wait(.25);  // Wait for a slightly longer duration before proceeding to the next iteration, as per the example code
}






core.setTimeRate(1);
core.waitFor("-0001-02-12T04:35:00", "local");
core.setTimeRate(0.1);



// Disable trails for planets
SolarSystem.setFlagTrails(false);



// showText("Date: " + core.getDate("local"), 5);





waitNextStep();






//
// STEP 4 - diurnal motion forward to evening of 01 June 2 BC
//


// real diurnal motion
for (i=0; i<25; i++)
{
    core.setDate("+30 minutes");
    core.wait(0.15);
}
showText("Date: " + core.getDate("local"), 5);

waitNextStep();

//
// - annual motion forward until 17 June 2 BC
//

LabelMgr.deleteAllLabels();
for (i=0; i<16; i++)
{
    core.setDate("+1 day");
    core.wait(0.15);
}
showText("Date: " + core.getDate("local"), 5);

waitNextStep();

//
//  - all constellations appear, slow diurnal motion
//

LabelMgr.deleteAllLabels();
ConstellationMgr.setFlagIsolateSelected(false);
ConstellationMgr.selectAllConstellations();
core.wait(0.1);
core.setTimeRate(200);
//core.setGuiVisible(true);

============================================================

**Reply by alex-w** [reply]
> Please can you take a look at the full script below - and specifically at the scripting in STEP 3 : // Loop to increment the date by two days at a time until "01 June 2 BC"

It's an easy - the comparison in `while()` is wrong, because both are strings. Please use something like: `while (core.getJDay() < 2460421)`, where `2460421` is a Julian Day for some date (please use Date and Time dialog to get the JD data ;) )

> The clock just does not run, and we don't know why 😧

The comparison of strings has other semantics in JavaScript

**Reply by ImmersiveTheatres** [reply]
OK awesome, thank you @alex-w. We now have the realtime date text changing in step with the annual motion! 😃

Yet it is displaying incorrect calendar data (down below please find the STEP 3 code as it currently stands).

Instead of the date progressing correctly from "12 August 3 BC" to "01 June 2 BC", the realtime date text is doing the following:

- "Bethlehem 12 August 3 BC" disappears
- "0 February NaN" appears
- 16 steps later, "10 February NaN" appears
- 16 steps later, "11 February NaN" appears
- 16 steps later, "12 February NaN" appears
- 16 steps later, "1 January NaN" appears
- 16 steps later, "2 January NaN" appears
- 16 steps later, "3 January NaN" appears
- 16 steps later, "4 January NaN" appears
- 16 steps later, "5 January NaN" appears
- 16 steps later, "6 January NaN" appears

Yet the date should be progressing  from "12 August 3 BC" to "01 June 2 BC". It is somehow stuck on January and February, with no year showing.

Please can you help us figure out why this is happening? (we are SO very close now!! 😊)

===================================

////////////////////////////////////////////////////////////////////////////////////////
//
//  STEP 3 - ALL CONSTELLATIONS FADE, ANNUAL MOTION FORWARD UNTIL MORNING OF 12 FEB 2 BC.
//


// DELETE THE PERSISTENT LABEL FROM STEP 1 BEFORE STARTING THE LOOP

LabelMgr.deleteLabel(bethlehemLabelID);



// FADE CONSTELLATIONS

ConstellationMgr.setFlagArt(false);
StelMovementMgr.lookZenith(); // does this cause the jump? why do we need this?
core.wait(0.1);



// waitNextStep();



// TURN ON CONSTELLATION LINES

ConstellationMgr.setFlagLines(true);



// waitNextStep();



// TURN ON TRAILS ONLY FOR JUPITER

SolarSystem.setFlagIsolatedTrails(true);
SolarSystem.setFlagTrails(true); 
core.selectObjectByName("Jupiter");



waitNextStep();



// ANNUAL MOTION TO 01 JUNE 2 BC

function formatDate(dateString) {
    var dateParts = dateString.split("-");
    var year = parseInt(dateParts[0]);
    var month = parseInt(dateParts[1]);
    var day = parseInt(dateParts[2]);

    var monthNames = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
    var formattedMonth = monthNames[month - 1];

    if (year < 0) {
        year = Math.abs(year) + 1 + " BC";  // Convert negative years to BC, adding 1 because there is no year 0
    }

    return day + " " + formattedMonth + " " + year;
}



// Existing code to create the initial date label
var labelID = LabelMgr.labelScreen("Bethlehem 12 August 3 BC", 160, 10, true, 30, "#ff0000");




// Loop to increment the date by two days at a time until "01 June 2 BC"

var targetJDay = 1720843.58976; // Julian Day for 01 June 2 BC

while (core.getJDay() < targetJDay) {
    core.setDate("+2 days"); // Increment the date by two days
    var currentDate = core.getDate("utc").toString().slice(0, 10); // Get the current date as a string in the format "YYYY-MM-DD"
    var formattedDate = formatDate(currentDate); // Format the date to the desired format "DD Month YYYY BC/AD"
    LabelMgr.deleteLabel(labelID); // Delete the old label
    labelID = LabelMgr.labelHorizon(formattedDate, 160, 10, true, 30, "#ff0000"); // Create a new label with the updated date and store its ID
    core.wait(1); // Wait for a short duration before proceeding to the next iteration
}



waitNextStep();

===================================

**Reply by alex-w** [reply]
I see in the code:
```
if (year < 0) {
    year = Math.abs(year) + 1 + " BC";  // Convert negative years to BC, adding 1 because there is no year 0
}
```
but apparently it should be
```
if (year <= 0) {
    year = Math.abs(year) + 1 + " BC";  // Convert negative years to BC, adding 1 because there is no year 0
}
```

**Reply by ImmersiveTheatres** [reply]
Thank you for your quick response @alex-w .

I've changed  (year < 0)  to  (year <=0) 

But this has made no difference. The annual motion runs, with the date changing - but with exactly the same problem as described above.

**Reply by alex-w** [reply]
@ImmersiveTheatres could you share a full script (via archive or repository)?

**Comment by ImmersiveTheatres** [comment]
https://github.com/ImmersiveTheatres/Stellarium/blob/main/Christmas_Star_10-11-23.ssc

**Reply by ImmersiveTheatres** [reply]
I hope I've done this correctly @alex-w : I've added the complete current script to my repository above.

**Reply by alex-w** [reply]
Please see https://github.com/ImmersiveTheatres/Stellarium/pull/1

**Comment by ImmersiveTheatres** [comment]
Woah! Excellent, the date text is now showing the correct months, in correct sequence, and the year too.

The only problem is August and September appear as "Undefined 3BC"

Then it correctly shows October, November, December, etc - increasing in two day increments - BUT the date "10" shows as "0" in every month:

1 February 2 BC
3 February 2 BC
5 February 2 BC
7 February 2 BC
**0 February 2 BC**
11 February 2 BC
13 February 2 BC
15 February 2 BC
17 February 2 BC
19 February 2 BC

Quick additional question: is it possible to add a line of code to disable the Moon ie. make it invisible?

Thank you for this! We're almost finished 😃✨

**Comment by ImmersiveTheatres** [comment]
I'm pleased to announce my new planetarium presentation, [The Christmas Star](https://immersive-theatres.com/the-christmas-star), created with a v22.3(v1.0) Stellarium script, was launched this week and has proved very successful.

The visuals look beautiful on the dome.

Thank you @alex-w for creating the first basic script, and to everyone above for your input. In the end the script required a LOT more blood, sweat and tears than I expected - yet the finished product has been worth it (you can see the Stellarium script running at the link above).
